### PR TITLE
https://github.com/paypal/squbs/issues/80

### PR DESCRIPTION
**Pull request by tutufool**
This patch:

1. simplify duplicate code for different type of messages handling like GET, POST, etc.
2. For each httpclient call, save 3 actor creations, they're:
HttpClientCallerActor, HttpClientPipelineTargetActor and HttpClientPipelineActor

### CODE REVIEW

File: squbs-httpclient/src/main/scala/org/squbs/httpclient/HttpClientCircuitBreaker.scala [hunk 1]
```diff
@@ -44,7 +44,7 @@ case class CircuitBreakerMetrics(var status: CircuitBreakerStatus,
 
 trait CircuitBreakerSupport{
 
-  def withCircuitBreaker(client: HttpClient, response: Future[HttpResponse])(implicit system: ActorSystem) = {
+  def withCircuitBreaker(client: HttpClient, response: => Future[HttpResponse])(implicit system: ActorSystem) = {
```
Comment (akara) [inline squbs-httpclient/src/main/scala/org/squbs/httpclient/HttpClientCircuitBreaker.scala]: Not clear why you choose to by-name the response?

Comment (tutufool) [inline squbs-httpclient/src/main/scala/org/squbs/httpclient/HttpClientCircuitBreaker.scala]: According to the definition:

  def withCircuitBreaker[T](body: ⇒ Future[T]): Future[T] = currentState.invoke(body)

This call is meant to protect the function that get the Future[T]

If we use by-value, which means we already call spray connector to get the Future[Response]

  def handle(client: HttpClient,
             pipeline: Try[HttpRequest => Future[HttpResponse]],
             httpRequest: HttpRequest)(implicit system: ActorSystem): Future[HttpResponse] = {
    implicit val ec = system.dispatcher
    httpClientLogger.debug("HttpRequest headers: " +
      httpRequest.headers.map { h => h.name + '=' + h.value } .mkString(", "))
    pipeline match {
      case Success(res) =>
        withCircuitBreaker(client, res(httpRequest))

----- we don't want to execute "res" function here, right?

**Comment by akara** [comment]
After working on the perf fork by @az-qbradley, I found this PR missed a test. Can you please take a look at ChunkedRequestSpec.scala line 89? That still starts the actor with the old signature and the file is not part of this PR. There may be other places. Thanks!

**Comment by tutufool** [comment]
The test failure is fixed.
which is caused by refactoring of PipelineProcessorActor, the constructor is changed.
I've sent email to Qian about how to fix this issues for similar cases in rocksqubs.

**Pull request merged**